### PR TITLE
Attempt to resolve some dangling pointer issues in Telemetry

### PR DESCRIPTION
**Pull request by joeltine**
b/298057575
b/298050585

Change-Id: I2f64ad74c667d258eb84105ed15b0d5f8f35105c

### CODE REVIEW

File: cobalt/browser/metrics/cobalt_metrics_log_uploader.cc [hunk 1]
```diff
@@ -67,8 +67,13 @@ void CobaltMetricsLogUploader::UploadLog(
       base::Base64UrlEncode(cobalt_uma_event.SerializeAsString(),
                             base::Base64UrlEncodePolicy::INCLUDE_PADDING,
                             &base64_encoded_proto);
-      upload_handler_->Run(h5vcc::H5vccMetricType::kH5vccMetricTypeCobaltUma,
-                           base64_encoded_proto);
+      // Check again that the upload handler is still valid. Was seeing race
+      // conditions where it was being destroyed while the proto encoding was
+      // happening above.
+      if (upload_handler_ != nullptr) {
```
Comment (jellefoks) [inline cobalt/browser/metrics/cobalt_metrics_log_uploader.cc:73]: I don't think this is complete.

When the application unloads or reloads or navigates the page, there is a time when the `H5vccMetrics` for the previous page is destroyed until a new `H5vccMetrics` is constructed. During that time `run_event_handler_callback_` and the object it holds is destroyed, and `upload_handler_` still has the address of that destroyed object.

So `upload_handler_` will dangle, but not be `nullptr`

Comment (joeltine) [reply cobalt/browser/metrics/cobalt_metrics_log_uploader.cc:73]: upload_handler_ is a unique_ptr upstream. See https://github.com/youtube/cobalt/blob/7eff4666ef9e3042b0841570e438218608624ea0/cobalt/h5vcc/h5vcc_metrics.cc#L36

It's my understanding that clang and gcc compilers set the underlying ptr to nullptr when destroying unique_ptrs: https://stackoverflow.com/questions/54237128/does-stdunique-ptr-set-its-underlying-pointer-to-nullptr-inside-its-destructor. 

```
On a side note, both clang (libc++) and gcc (libstdc++) sets the pointer to nullptr when destroying a std::unique_ptr, but here is gcc destructor:
```

Comment (jellefoks) [reply cobalt/browser/metrics/cobalt_metrics_log_uploader.cc:73]: Iy's a bare pointer here, and it's here where it's dangling when that unique_ptr gets destroyed.

Comment (joeltine) [reply cobalt/browser/metrics/cobalt_metrics_log_uploader.cc:73]: Ah, right, for some reason I was thinking this was a reference to a reference.

I'm struggling thinking how best to handle this. Any recommendations on how to work around this? How do I do this safely when the JS callback object can be destroyed outside this class's control?

Comment (joeltine) [reply cobalt/browser/metrics/cobalt_metrics_log_uploader.cc:73]: K. PTAL. Per our conversation this morning, I've added a destructor `~H5vccMetrics` that sets the upload callback to `nullptr`. A mutex wasn't necessary as the CobaltMetricsServiceManager uses PostTask to ensure its called on the same thread every time.

File: cobalt/h5vcc/h5vcc_metrics.cc [hunk 2]
```diff
@@ -25,6 +25,16 @@
 namespace cobalt {
 namespace h5vcc {
 
+H5vccMetrics::~H5vccMetrics() {
+  if (browser::metrics::CobaltMetricsServicesManager::GetInstance() !=
+      nullptr) {
+    // We need to let the metrics manager know not to call the upload callback
+    // any longer, otherwise it could crash.
+    browser::metrics::CobaltMetricsServicesManager::GetInstance()
+        ->SetOnUploadHandler(nullptr);
```
Comment (jellefoks) [inline cobalt/h5vcc/h5vcc_metrics.cc:34]: Being paranoid here, this object gets destroyed when the JS engine for a page with this object is destroyed, but I'm not certain that it's guaranteed that will happen before a new page is constructed. A new H5vccMetrics may be constructed before this one is destroyed. Perhaps the `SetOnUploadHandler(nullptr)` can be replaced with a `RemoveOnUploadHandler(this)` which then only sets it to null of the handler was the pointer passed in.

Comment (joeltine) [reply cobalt/h5vcc/h5vcc_metrics.cc]: Done

**Comment by joeltine** [comment]
@jellefoks friendly ping, this review is now out of SLO

**Review by jellefoks — APPROVED** [review]

**Pull request merged**